src/webpack/plugins/BasePlugin.ts

```diff
@@ -0,0 +1,179 @@
+import * as fs from 'fs';
+import * as path from 'path';
+
+import * as webpack from 'webpack';
+import * as mkdirp from 'mkdirp';
+
+interface BaseOptions {
+    outputPath: string;
+    input: string|string[];
+}
+
+export type PluginSpec<T> = {
+    [K in keyof T]: T[K]|((value: T[K]) => any);
+}
+
+export type PluginOptions<T> = T & BaseOptions;
+
+export abstract class BasePlugin<T> {
+    /**
+     * Supported options:
+     * - outputPath: string
+     *   Where to write any output files to, either a filename or a directory.
+     *   If a directory, filenames will be appended based on the input file with the extension changed.
+     * - input: string|string[]
+     *   The JSON Schema file(s) to process using this plugin.
+     *   If passing multiple files, then 'outputPath' MUST be a directory rather than an absolute filename.
+     *
+     * @param name The unique name of this plugin. Will appear in webpack output.
+     * @param outputExtension The default file extension to use when writing files from this plugin. If only writing a single file, can be overridden by using an absolute path as outputPath option
+     * @param options Object containing options passed-in from webpack config. See comments on BasePlugin and derived classes for list of supported options.
+     */
```

Comment:
Should this be on the constructor? Can you move the `outputPath`/`input` comments to the `BaseOptions` interface?

Reply:
That makes sense.  I will do that.

Reply:
Fixed